### PR TITLE
Propagate Headers and Context through to ScriptService

### LINKED ISSUE #12891: Requests do not always propagate headers/context when retrieving indexed scripts
In #11060 (issue #10979), the request headers and context of the originating request were copied to all sub-requests. In doing some testing on master, I noticed that in the `ScriptService#getScriptFromIndex` the headers and context were not being copied due to the lack of a `SearchContext`. To see what calls were affected, I added an assert to check if the context was null. When doing so the following tests trip the assertion:

```
- org.elasticsearch.search.aggregations.pipeline.BucketScriptIT.indexedScript
- org.elasticsearch.search.aggregations.metrics.ScriptedMetricIT.testInitMapCombineReduce_withParams_Indexed
- org.elasticsearch.search.aggregations.pipeline.BucketSelectorIT.indexedScript
- org.elasticsearch.validate.RenderSearchTemplateIT.indexedTemplate
```

There were also two rest test failures for the Render Search Template API. For the render search template api, we need a way to pass in the headers/context in a way that isn't a search context because it is not a search request.

For the other failures, it appears that the subsequent requests happen after the SearchContext has been cleared/released.

### PR DESCRIPTION
**Pull request by colings86**
At the moment if an index script is used in a request, the spawned request to get the indexed script from the `.scripts` index does not get the headers and context copied to it from the original request. This change makes the calls to the `ScriptService` pass in a `HasContextAndHeaders` object that can provide the headers and context. For the `search()` method the context and headers are retrieved from `SearchContext.current()`.

Closes #12891

### CODE REVIEW

**Comment by spinscale** [comment]
can we add tests to `ContextAndHeaderTransportIT` for testing this?

**Comment by colings86** [comment]
@spinscale I added a couple of tests to check that the headers and context are propagated for the TemplateQuery and for the reduce phase of aggregations.

**Comment by spinscale** [comment]
overall this looks good to me, but

* I couldnt run `mvn test` due to forbidden API errors, is it me?
* Can we add a test for the phrase suggesters as it has been changed?
* Thinking if we can reuse the `DelegatingHasContextAndHeaders` in our `SearchContext` (might be tricky with the inheritance though)

**Comment by colings86** [comment]
@spinscale I have implemented the suggested changes and rebased onto master (without squashing commits), are you able to take another look?

File: core/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 1]
```diff
@@ -67,7 +69,7 @@
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-public abstract class SearchContext implements Releasable, HasContextAndHeaders {
+public abstract class SearchContext extends DelegatingHasContextAndHeaders implements Releasable {
```
Comment (spinscale) [inline core/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: with this change in place, can you try to remove the `HasContextAndHeaders` delegation methods from `PercolateContext` and `FilteredSearchContext`?

**Comment by spinscale** [comment]
left one last comment, I am not sure if you can remove all the methods (having the asserts in the percolate context seems useful to me), but maybe you can remove some code...

apart from that LGTM

**Comment by colings86** [comment]
I was able to remove the methods from FilteredSearchContext but I kept them in PercolateContext as I think the asserts are important to have.

**Pull request merged**